File: src/simulation/culture/group/cultureaspect/RitualSystem.java
package simulation.culture.group.cultureaspect;

import shmp.random.RandomProbabilitiesKt;
import simulation.Controller;
import simulation.culture.group.centers.Group;
import simulation.culture.group.reason.Reason;
import simulation.culture.group.request.Request;

import java.util.Collection;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

public class RitualSystem implements CultureAspect {
    private Set<Ritual> rituals;
    private Reason reason;

    public RitualSystem(Group group, Collection<Ritual> rituals, Reason reason) {
        this.rituals = new HashSet<>(rituals);
        this.reason = reason;
    }

    public Reason getReason() {
        return reason;
    }

    public Set<Ritual> getRituals() {
        return rituals;
    }

    public void addRitual(Ritual ritual) {
        rituals.add(ritual);
    }

    @Override
    public Request getRequest(Group group) {
        return null;
    }

    @Override
    public void use(Group group) {
        rituals.forEach(r -> r.use(group));
        if (!RandomProbabilitiesKt.testProbability(
                Controller.session.groupCollapsedAspectUpdate,
                Controller.session.random)
        ) {
            return;
        }
        Ritual ritual = ConstructCultureAspectKt.constructRitual(
                reason,
                group,
                Controller.session.random
        );
        if (ritual != null) {
            rituals.add(ritual);
        }
    }

    @Override
    public RitualSystem adopt(Group group) {
        return new RitualSystem(group, rituals.stream()
                .map(r -> (Ritual) r.adopt(group)).collect(Collectors.toList()),
                reason);
    }

    @Override
    public String toString() {
        return String.format("Ritual system for %s", reason);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        RitualSystem that = (RitualSystem) o;
        return Objects.equals(rituals, that.rituals);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rituals);
    }

    @Override
    public void die(Group group) {
        rituals.forEach(r -> r.die(group));
    }
}

File: src/simulation/culture/group/cultureaspect/RitualSystem.kt
package simulation.culture.group.cultureaspect

import shmp.random.testProbability
import simulation.Controller.*
import simulation.culture.group.centers.Group
import simulation.culture.group.reason.Reason
import simulation.culture.group.request.Request
import java.util.*

class RitualSystem constructor(rituals: Collection<Ritual>, val reason: Reason) : CultureAspect {
    private val _rituals: MutableSet<Ritual>

    val rituals: Set<Ritual>
        get() = _rituals

    init {
        this._rituals = HashSet(rituals)
    }

    fun addRitual(ritual: Ritual) {
        _rituals.add(ritual)
    }

    override fun getRequest(group: Group): Request? = null

    override fun use(group: Group) {
        _rituals.forEach { it.use(group) }
        if (!testProbability(session.groupCollapsedAspectUpdate, session.random))
            return
        val ritual = constructRitual(reason, group, session.random)
        if (ritual != null)
            _rituals.add(ritual)
    }

    override fun adopt(group: Group): RitualSystem? {
        val newRituals = _rituals.map { it.adopt(group) }
        if (newRituals.any { it == null }) return null
        return RitualSystem(_rituals, reason)
    }

    override fun toString() = "Ritual system for $reason"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as RitualSystem
        return _rituals == that._rituals
    }

    override fun hashCode() = Objects.hash(_rituals)

    override fun die(group: Group) = _rituals.forEach { it.die(group) }
}